### PR TITLE
Add a Why Migrate? guide

### PR DESCRIPTION
**Pull request by casperisfine**
I didn't want to open the can of worm of comparison tables with other servers, but I've received some comments that suggest it may be a good idea to more clearly state the expected tradeoffs to manage expectations.

### CODE REVIEW

File: docs/WHY_MIGRATE.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+If a bug cause Ruby to hand, the monitor process will SIGKILL the worker and the capacity will be
+reclaimed.
+
+This makes Pitchfork more resilient to some classees of bugs.
+
+#### Thread Safety
+
+Pitchfork doesn't require applications to be thread-safe. That is probably the worst reason
+to migrate thought.
```
Comment (dzunk) [inline docs/WHY_MIGRATE.md:80]: "thought" => "though". If this is a bad reason, should we even mention it?

File: docs/WHY_MIGRATE.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Since Pitchfork workers have their own address space and only process one request at a time
+it makes it much harder for one faulty request to impact another.
+
+Even if a bug cause Ruby to crash, only the request that triggered the bug will be impacted.
+
+If a bug cause Ruby to hand, the monitor process will SIGKILL the worker and the capacity will be
+reclaimed.
+
+This makes Pitchfork more resilient to some classees of bugs.
```
Comment (dzunk) [inline docs/WHY_MIGRATE.md:75]: "classees" => "classes"

File: docs/WHY_MIGRATE.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+This can significantly improve tail latency at the expense of throughput.
+
+#### Resiliency and Isolation
+
+Since Pitchfork workers have their own address space and only process one request at a time
+it makes it much harder for one faulty request to impact another.
+
+Even if a bug cause Ruby to crash, only the request that triggered the bug will be impacted.
+
+If a bug cause Ruby to hand, the monitor process will SIGKILL the worker and the capacity will be
```
Comment (dzunk) [inline docs/WHY_MIGRATE.md:72]: "cause Ruby to hand" => "causes Ruby to hang"

File: docs/WHY_MIGRATE.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+[it allows to periodically trigger garbage collection when the worker isn't processing any request](https://shopify.engineering/adventures-in-garbage-collection).
+
+This can significantly improve tail latency at the expense of throughput.
+
+#### Resiliency and Isolation
+
+Since Pitchfork workers have their own address space and only process one request at a time
+it makes it much harder for one faulty request to impact another.
+
+Even if a bug cause Ruby to crash, only the request that triggered the bug will be impacted.
```
Comment (dzunk) [inline docs/WHY_MIGRATE.md:70]: "cause" => "causes"

File: docs/WHY_MIGRATE.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+#### Memory Usage
+
+Without reforking enabled Pitchfork will without a doubt use more memory than threaded Puma.
+
+With reforking enabled, results will vary based on the application profile and the number of Puma threads,
+but should be in the same ballpark, sometimes better, but likely worse, this depends on many variables and
+can't really be predicted.
+
+However be warned that making an application fork safe can be non-trivial,
```
Comment (dzunk) [inline docs/WHY_MIGRATE.md:92]: Thoughts on adding more context here? I'm not sure where the difficulty lies in doing this.

**Review by dzunk — COMMENTED** [review]
Sensible addition, helpful context. Sadly at work we use Passenger which is a whole different can of worms 😄

File: docs/WHY_MIGRATE.md (hunk 5 above)
Comment (casperisfine) [reply docs/WHY_MIGRATE.md]: I added a link to the fork safety guide, thanks.

File: docs/WHY_MIGRATE.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+#### Reduced Features
+
+While Pitchfork started as a fork of Unicorn, many features such as daemonization,
+pid file management, hot reload have been stripped.
+
+Pitchfork only kept features that makes sense in a containerized world.
+
+## Coming from Puma
+
+Generally speaking, compated to (threaded) Puma, Pitchfork may offer better latency and isolation at the expense of throughput.
```
Comment (jordelver) [inline docs/WHY_MIGRATE.md:44]:
```suggestion
Generally speaking, compared to (threaded) Puma, Pitchfork may offer better latency and isolation at the expense of throughput.
```

(I see this is still draft, so ignore this if not appropriate :)

**Pull request merged**